README.md
<p align="center">
  <img alt="Who Is Sciper" src="https://raw.githubusercontent.com/innovativeinnovation/who-is-sciper/master/docs/readme/readme-logo.png">
</p>

<p align="center">
  Find out who a person is given a sciper.
</p>

<p align="center">
  <a href="https://github.com/innovativeinnovation/who-is-sciper/actions">
    <img alt="Build Status" src="https://github.com/innovativeinnovation/who-is-sciper/workflows/Build/badge.svg?branch=master">
  </a>
  <a href="https://raw.githubusercontent.com/innovativeinnovation/who-is-sciper/master/LICENSE">
    <img alt="Apache License 2.0" src="https://img.shields.io/badge/license-Apache%202.0-blue.svg">
  </a>
  <a href='https://www.npmjs.com/package/who-is-sciper'>
    <img alt="NPM Version" src="https://img.shields.io/npm/v/who-is-sciper.svg" />
  </a>
</p>

---

Install
-------

Install this globally and you'll have access to the `who-is-sciper` command
anywhere on your system.

```bash
npm i who-is-sciper -g
```

Usage
-----

```console
who-is-sciper
Usage: who-is-sciper <sciper>

Commands:
  sciper  6-digit unique EPFL identification number.

Options:
  -l, --language  Show informations in "en" or "fr"       [string]
  -h, --help      Show help                               [boolean]
  -v, --version   Show version number                     [boolean]

Examples:
  who-is-sciper 128871
  who-is-sciper 278890 -l fr
```

Screenshot
----------

![command line screenshot](https://raw.githubusercontent.com/innovativeinnovation/who-is-sciper/master/docs/readme/screenshot.png)

See also
--------

* [epfl-people-api](https://github.com/innovativeinnovation/epfl-people-api)

Contributing
------------

Contributions are always welcome.

See [Contributing](CONTRIBUTING.md).

Developer
---------

* [William Belle](https://github.com/williambelle)

License
-------

Apache License 2.0

Original work (c) ECOLE POLYTECHNIQUE FEDERALE DE LAUSANNE, Switzerland, VPSI, 2017-2018.  
Modified work (c) William Belle, 2018-2023.

See the [LICENSE](LICENSE) file for more details.


CHANGELOG.md
Changelog
=========

### v7.0.0 / 2023-11-07

  - Drop support for Node < 16
  - Update dependencies
  - Update Github Actions checkout to v4
  - Update Github Actions setup-node to v4
  - Test against Node.js 21

### v6.0.1 / 2023-07-14

  - Test against Node.js 20
  - Update dependencies

### v6.0.0 / 2022-12-15

  - Drop support for Node < 14
  - Test against Node.js 19
  - Update dependencies

### v5.1.1 / 2022-08-12

  - Test against Node.js 18
  - Update dependencies

### v5.1.0 / 2022-04-05

  - Add unit name
  - Remove address logic (old API code)
  - Fix tests

### v5.0.2 / 2021-11-26

  - Migrate to new API
  - Test against Node.js 17

### v5.0.1 / 2021-08-19

  - Update dependencies

### v5.0.0 / 2021-06-03

  - Drop support for Node.js < 12
  - Test against Node.js 16
  - Update dependencies

### v4.0.3 / 2021-03-17

  - Update and clean dependencies
  - Migrate from Travis CI to GitHub Actions
  - Remove badge dependencies

### v4.0.2 / 2020-10-27

  - Update dependencies
  - Test against Node.js 15
  - Migrate travis-ci.org to .com

### v4.0.1 / 2020-05-26

  - Update dependencies
  - Test against Node.js 14

### v4.0.0 / 2020-01-05

  - Drop support for Node < 10
  - Update dependencies

### v3.1.2 / 2019-12-13

  - Update dependencies
  - Test against Node.js 13

### v3.1.1 / 2019-10-04

  - Update dependencies

### v3.1.0 / 2019-07-13

  - Display profile
  - Update dependencies

### v3.0.5 / 2019-04-30

  - Improve documentation
  - Update dependencies
  - Update screenshot
  - Test against Node.js 12

### v3.0.4 / 2019-02-21

  - Update dependencies
  - Update screenshot

### v3.0.3 / 2019-01-24

  - Update dependencies
  - Test against Node.js 11

### v3.0.2 / 2018-10-24

  - Update copyright (transfer from EPFL DevRun)

### v3.0.1 / 2018-09-20

  - Update dependencies

### v3.0.0 / 2018-08-09

  - Drop support for Node < 8
  - Update dependencies
  - Update cli example

### v2.0.0 / 2018-06-15

  - Drop support for Node < 6
  - Rewrite code
  - Move to david-dm (watching deps)
  - Update dependencies

### v1.1.0 / 2018-03-28

  - Move to eslint
  - Update dependencies
  - Update license copyright date
  - Disable `package-lock.json`

### v1.0.1 / 2018-01-25

  - Update dependencies

### v1.0.0 / 2017-11-24

  - Update dependencies

### v0.1.3 / 2017-09-27

  - Update dependencies
  - Simplify people url

### v0.1.2 / 2017-08-04

  - Move translations terms in cli

### v0.1.1 / 2017-08-04

  - Fix missing file in package.json

### v0.1.0 / 2017-08-04

  - Show accreditations of a person
  - Add option `-l` to show informations in 'en' or 'fr'

### v0.0.2 / 2017-07-27

  - Add better error message
  - Fix undefined email
  - Update dependencies

### v0.0.1 / 2017-07-18

  - First version, released on an unsuspecting world.


test/test.js
/*
 * Original work (c) ECOLE POLYTECHNIQUE FEDERALE DE LAUSANNE, Switzerland,
 * VPSI, 2017-2018.
 * Modified work (c) William Belle, 2018-2023.
 * See the LICENSE file for more details.
 */

require('chai').should();
const version = require('../package').version;

describe('who-is-sciper cli', function () {
  this.timeout(15000);
  let cliOption = '-v';
  let response;

  beforeEach((done) => {
    const execFile = require('child_process').execFile;
    execFile('./src/cli.js', [cliOption], (error, stdout) => {
      if (error) {
        throw error;
      }
      response = stdout;
      done();
    });
  });

  it('should match version with option -v', () => {
    response.should.equal(version + '\n');
    cliOption = '100000';
  });

  it('should return error with sciper 100000', () => {
    response.should.match(/Sciper does not exist/);
    cliOption = '69';
  });

  it('should return error with sciper 69', () => {
    response.should.match(/Expected a sciper/);
    cliOption = '128871';
  });

  it('should match person with sciper 128871', () => {
    response.should.match(/Duratti/);
    response.should.match(/Lindo/);
    response.should.match(/INN 0/);
    response.should.match(/FSD/);
    response.should.match(/Development/);
  });
});


src/cli.js
#!/usr/bin/env node

/*
 * Original work (c) ECOLE POLYTECHNIQUE FEDERALE DE LAUSANNE, Switzerland,
 * VPSI, 2017-2018.
 * Modified work (c) William Belle, 2018-2023.
 * See the LICENSE file for more details.
 */

const epflPeopleApi = require('epfl-people-api');
const detectGender = require('gender-detection');
const chalk = require('chalk');
const yargs = require('yargs')

  // Sciper
  .command('sciper', '6-digit unique EPFL identification number.')
  .required(1, 'sciper is required')

  // Language
  .option('l', {
    alias: 'language',
    describe: 'Show informations in "en" or "fr"',
    requiresArg: true,
    type: 'string'
  })

  // Version
  .alias('v', 'version')

  // Help
  .help('h')
  .alias('h', 'help')
  .usage('Usage: $0 <sciper>')
  .example('$0 128871')
  .example('$0 278890 -l fr');

const argv = yargs.argv;
const sciper = argv._[0];
let locale = 'en';

const i18n = {
  en: {
    name: 'Name',
    email: 'Email',
    profile: 'Profile',
    position: 'Position',
    unit: 'In unit',
    office: 'Office',
    phone: 'Phone',
    see: 'See'
  },
  fr: {
    name: 'Nom',
    email: 'Email',
    profile: 'Profil',
    position: 'Position',
    unit: 'Dans l\'unité',
    office: 'Bureau',
    phone: 'Téléphone',
    see: 'Voir'
  }
};

const MAX_LINE_LENGTH = { en: 10, fr: 14 };

const createText = (key, maxLength) => {
  let text = '';
  if (key) {
    text = i18n[locale][key] + ':';
  }
  for (let i = text.length; i < maxLength; i++) {
    text += ' ';
  }
  return text;
};

const aggregateInfos = (person) => {
  let infos = '';
  infos += createText('name', MAX_LINE_LENGTH[locale]) + person.firstname +
    ' ' + person.name + '\n';
  if (person.email) {
    infos += createText('email', MAX_LINE_LENGTH[locale]) + person.email +
      '\n';
  }
  if (person.profile) {
    infos += createText('profile', MAX_LINE_LENGTH[locale]) + person.profile +
      '\n';
  }
  for (let i = 0; i < person.accreds.length; i++) {
    infos += '\n' + createText('position', MAX_LINE_LENGTH[locale]) +
      person.accreds[i].position + '\n';
    infos += createText('unit', MAX_LINE_LENGTH[locale]) +
      person.accreds[i].acronym + ' (' + person.accreds[i].name + ')' + '\n';

    if (person.accreds[i].officeList &&
        person.accreds[i].officeList.length > 0) {
      infos += createText('office', MAX_LINE_LENGTH[locale]) +
        person.accreds[i].officeList.join(', ') + '\n';
    }
    if (person.accreds[i].phoneList &&
        person.accreds[i].phoneList.length > 0) {
      infos += createText('phone', MAX_LINE_LENGTH[locale]) +
        person.accreds[i].phoneList.join(', ') + '\n';
    }
  }
  return infos;
};

const put = (firstname, sciper, infos) => {
  infos += '\n' + i18n[locale].see +
    ' https://people.epfl.ch/' + sciper;
  const gender = detectGender.detect(firstname);
  if (gender === 'female') {
    try {
      console.log(chalk.hex('#FD4286')(infos));
    } catch (err) {
      console.log(infos);
    }
  } else if (gender === 'male') {
    try {
      console.log(chalk.hex('#0484FF')(infos));
    } catch (err) {
      console.log(infos);
    }
  } else {
    console.log(infos);
  }
};

const displayFull = (person) => {
  put(
    person.firstname,
    person.sciper,
    aggregateInfos(person)
  );
};

if (argv.l && argv.l === 'fr') {
  locale = 'fr';
}

epflPeopleApi.findBySciper(sciper, locale).then((person) => {
  displayFull(person);
}).catch((err) => console.log(err.message));
